pkg/azure/provider.go

```diff
@@ -553,44 +555,45 @@ func RedactClientID(sensitiveString string) string {
 	return r.ReplaceAllString(sensitiveString, "$1##### REDACTED #####$3")
 }
 
-// getCertAndPrivKeyInPEMFormat returns the certificate and private key to be
-// written to file
-// cert and private key are returned when object type is "secret"
-func getCertAndPrivKeyInPEMFormat(value string) (string, error) {
-	return decodePKCS12(value, true, true)
-}
-
 // decodePkcs12 decodes a PKCS#12 client certificate by extracting the public certificate and
 // the private key
-func decodePKCS12(value string, getKey, getCert bool) (content string, err error) {
+func decodePKCS12(value string) (content string, err error) {
 	pfxRaw, err := base64.StdEncoding.DecodeString(value)
 	if err != nil {
 		return "", err
 	}
-	key, cert, err := pkcs12.Decode(pfxRaw, "")
+	// using ToPEM to extract more than one certificate and key in pfxData
+	pemBlock, err := pkcs12.ToPEM(pfxRaw, "")
 	if err != nil {
 		return "", err
 	}
-	keyX509, err := x509.MarshalPKCS8PrivateKey(key)
-	if err != nil {
-		return "", err
-	}
-	var pemData []byte
-	if getKey {
-		keyBlock := &pem.Block{
-			Type:  "PRIVATE KEY",
-			Bytes: keyX509,
-		}
-		pemData = append(pemData, pem.EncodeToMemory(keyBlock)...)
-	}
 
-	if getCert {
-		certBlock := &pem.Block{
-			Type:  "CERTIFICATE",
-			Bytes: cert.Raw,
+	var pemKeyData, pemCertData, pemData []byte
+	for _, block := range pemBlock {
+		// PEM block encoded form contains the headers
+		//    -----BEGIN Type-----
+		//    Headers
+		//    base64-encoded Bytes
+		//    -----END Type-----
+		// Setting headers to nil to ensure no headers included in the encoded block
+		block.Headers = make(map[string]string)
+		if block.Type == certificateType {
+			pemCertData = append(pemCertData, pem.EncodeToMemory(block)...)
+		} else {
+			key, err := parsePrivateKey(block.Bytes)
+			if err != nil {
+				return "", err
+			}
+			// converting to pkcs8 private key as ToPEM uses pkcs1
+			block.Bytes, err = x509.MarshalPKCS8PrivateKey(key)
```

Comment:
can you add some comments for why we are converting to pkcs8?

Reply:
Added a comment